easy-ui-react/src/DataGrid/DataGrid.tsx

```diff
@@ -0,0 +1,222 @@
+    setExpandedKey(expandedKeyFromUser);
+  }
+
+  const columns = useProcessedColumns(props);
+  const rows = useProcessedRows(props, expandedKey);
+
+  const context = useMemo(() => {
+    return { expandedKey, setExpandedKey };
+  }, [expandedKey]);
+
+  return (
+    <DataGridContext.Provider value={context}>
+      <Table {...props}>
+        <TableHeader columns={columns}>
+          {(column) => (
+            <Column
+              isRowHeader={column.key === rowHeaderColumnKey}
+              allowsSorting={
+                column.key === EXPAND_COLUMN_KEY ||
+                column.key === ACTIONS_COLUMN_KEY
+                  ? false
+                  : columnKeysAllowingSort.includes(column.key)
+              }
+            >
+              {column.key === EXPAND_COLUMN_KEY ? (
+                <VisuallyHiddenCellContent>
+                  Expand row
+                </VisuallyHiddenCellContent>
+              ) : column.key === ACTIONS_COLUMN_KEY ? (
+                <VisuallyHiddenCellContent>
+                  Row actions
+                </VisuallyHiddenCellContent>
+              ) : (
+                renderColumnCell(column)
+              )}
+            </Column>
+          )}
+        </TableHeader>
+        <TableBody items={rows}>
+          {(row) => (
+            <Row>
+              {(columnKey) => (
+                <Cell>
+                  {columnKey === EXPAND_COLUMN_KEY ? (
+                    <ExpandCellContent
+                      isExpanded={row.key === expandedKey}
+                      toggleExpanded={() => toggleExpandedRow(row.key)}
+                    />
+                  ) : columnKey === ACTIONS_COLUMN_KEY && rowActions ? (
+                    <ActionsCellContent rowActions={rowActions(row.key)} />
+                  ) : (
+                    renderRowCell(
+                      row[columnKey as keyof typeof row],
+                      columnKey,
+                      row,
+                    )
+                  )}
+                </Cell>
+              )}
+            </Row>
+          )}
+        </TableBody>
+      </Table>
+    </DataGridContext.Provider>
+  );
+}
+
+/**
+ * Modifies the passed in columns to include support for Easy UI requirements.
+ * This is done before being passed into React Stately's Column interface.
+ *
+ * @param props data grid props
+ * @returns processed columns
+ */
+function useProcessedColumns<C extends ColumnType>(
+  props: Pick<DataGridProps<C>, "renderExpandedRow" | "columns" | "rowActions">,
+) {
+  const { renderExpandedRow, columns, rowActions } = props;
+  const hasExpandableRows = Boolean(renderExpandedRow);
+  const hasRowActions = Boolean(rowActions);
+  return useMemo(() => {
+    let c = columns;
+    if (hasExpandableRows) {
+      c = [{ key: EXPAND_COLUMN_KEY } as C, ...c];
+    }
+    if (hasRowActions) {
+      c = [...c, { key: ACTIONS_COLUMN_KEY } as C];
+    }
+    return c;
```

Comment:
early returns wouldn't allow for both conditions to process unfortunately

Reply:
Oh yeah duh, I didn't even think about the fact that something could have both of those be true. No need to implement this change, but I tend to prefer optionally including items in an array instead of mutating it. So something like this:
```
return [
  ...(hasExpandableRows ? [{ key: EXPAND_COLUMN_KEY } as C] : []),
  ...columns,
  ...(hasRowActions ? [{ key: ACTIONS_COLUMN_KEY } as C] : []),
]
```
The reason that I prefer this format is because it indicates more clearly the intention of the code. The code that you have right now reads like this (at least to me): Create a new array from the value of columns, if it `hasExpandableRows` then create a new array that starts with `{ key: EXPAND_COLUMN_KEY }` and contains the rest of the old array and reassign the value of the variable that was just created, if it `hasRowActions` then create a new array that starts with `{ key: ACTIONS_COLUMN_KEY }` and contains the rest of the old array and reassign the value of the variable that was just created, finally return the newly created array.

Alternatively, my proposed code reads like: create and return a new array, set the first item in the returned array as `{ key: EXPAND_COLUMN_KEY }` if `hasExpandableRows` is true, include the items from columns in this array, set the last item in the returned array as `{ key: ACTIONS_COLUMN_KEY }` if `hasRowActions` is true.

Besides reading a little more clearly, it also minimizes places where things can go wrong. For instance when we do something like what is currently in the code we risk `if (hasRowActions)` breaking something that was done earlier (say we forget to spread `c` inside of the `if` statement). 

Like I said, no need to change this. I just wanted to point out a way that I think is a bit easier to read/understand in order to do something like this.